### ISSUE TITLE
standalone-verify fails with Error verifying signature: Unexpected GPG signature count 0 when no tty is allocated

### ISSUE THREAD

**Issue by hamburgerbun** [containers/skopeo#1218]
skopeo version 1.1.0
OS: MacOS 10.14.6, Docker desktop 3.1.0 (51484), ubuntu:20.10 container base

I'm using a toolset to use skopeo for creating signatures for containers, and before I upload them to a signature webstore, I wanted to verify them with 
```
skopeo standalone-verify ./manifest.json <full_container_tag> <GPG fingerprint> ./signature-file
```

this works perfectly when i do it manually within the docker container, i.e. docker run -it --rm /bin/bash, and then i manually type out the above after dealing with synaptic and importing the relevant gpg key. 

this also works perfectly when I execute this in the context of a bash script from me doing docker run -it on my mac terminal, like docker run -it --rm signingcontainer:v1 /sign_me.sh arg1 arg2 arg3, and then the script does whatever i did manually with the signing, the verify, and upload to webstore. 

this however doesn't work when i am trying to run that same docker run command above within a gradle task after dropping -it (because we no longer have a chardev input), so just docker run --rm /sign_me.sh args... my standalone-verify call errors out with
```
time="2021-03-03T23:21:06Z" level=fatal msg="Error verifying signature: Unexpected GPG signature count 0"
```

finally, this works fine in my gradle task when i don't link up stdin, and just have the pseudo tty allocated ala docker run -t --rm, so presumably, it's the pseudo-tty here that makes the magic happen.

Why does the tty make a difference for standalone-verify?

**Comment by vrothberg** [comment]
Thanks for reaching out.  Can you try with the latest Skopeo?

**Comment by vrothberg** [comment]
I expect it to fail with v1.2.1 as well since it looks like a GPG behavior.

**Comment by mtrmac** [comment]
What does `gpg -d ./signature-file` do in that environment?

**Comment by mtrmac** [comment]
Also, is that _exactly_ the same signature+private key? That code paths seems to be a “GPG successfully verified the input, but there were no signatures” case.

**Comment by hamburgerbun** [comment]
if i run it in the docker container manually or include it in my script with -t, i get exit code 0 and the following output (i replaced a bunch of the strings lol 
```
{"critical":{"identity":{"docker-reference":"<container full tag>"},"image":{"docker-manifest-digest":"sha256:<sha hash stuff>"},"type":"atomic container signature"},"optional":{"creator":"","timestamp":1614811395}}gpg: Signature made Wed Mar  3 22:43:15 2021 UTC
gpg:                using RSA key 123412341234123
gpg: Good signature from "test signing" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 1234 1234 1234 1234 1234 1234 1234 1234
```

when i run it in the container with to tty, i see this error instead
```
gpg: block_filter: 1st length byte missing
gpg: cannot open '/dev/tty': No such device or address
```

it's the same private key every time, the generated signature i don't think so since there's a timestamp in the content.

**Comment by davidhay1969** [comment]
@hamburgerbun quick question, did you get to the bottom of this, as I'm seeing something similar, albeit with a completely different setup ( running `skopeo` inside a Fedora `34` QEMU VM under the aegis of [Kata Containers](https://github.com/kata-containers/kata-containers/releases) ) ?

For reference I'm running `skopeo` `1.2.2` as per: -

`skopeo --version`

```text
skopeo version 1.2.2
```

**Comment by mtrmac** [comment]
So far random searches suggest that the input data is probably corrupt. But I don’t know whether “input data” specifically means the signature, the public key, or perhaps even some communication with a key agent. Or it just could be a red herring…

**Comment by davidhay1969** [comment]
Hey @mtrmac yeah, that's what I was thinking but .... like @hamburgerbun says, it only borks when running `skopeo` in "headless"  mode ...

In my particular setup, I'm invoking `skopeo standalone-verify` via the `kata-agent` which is written in Rust, and runs inside a Fedora `34` virtual machine ( there's more not-terribly-relevant-here backstory ... ).

TL;DR; `skopeo` runs as the result of a remote API call.

When I run `standalone-verify` in this headless mode, I see: -

```text
Error verifying signature: Unexpected GPG signature count 0 
```

However, when I run it interactively ( inside the Fedora VM ) -

`skopeo standalone-verify /tmp/image/manifest.json us.icr.io/foobar/hello_world_nginx_june_2021:latest 6C6428941720EDD47AFB8DEE284F019C785EB0109 /tmp/image/signature-1 `

all is well: -

```text
Signature verified, digest sha256:499e404dcc16210719b6b61aae27f7a3c3576daf49c4eacabfbce37d0c593f17
```

Similarly, if I run the same command via a Bash script, again inside the Fedora VM, I see the same positive result.

Right now, I'm looking at `skopeo --version` - the version that "comes with" Fedora is `1.2.2` whereas I *think* that the problem doesn't recur with `1.3.0` ...

Just tinkering further, and will report back .....

**Comment by mtrmac** [comment]
If the `gpg -d` command suggested above differs in behavior under the same pair of conditions, that’s where the debugging focus needs to be. I can’t recommend a specific process for debugging the behavior of the GPG implementation, I’m afraid, I have only needed to do so on a few rare occasions.

**Comment by davidhay1969** [comment]
No worries, will keep on pluggin' away - and will report back!

**Comment by davidhay1969** [comment]
I *think* I may have nailed it .... doing some more investigation, but I suspect that the problem may be that the GPG keyring is, for my setup, in a very non-standard environment, meaning that I need to explicitly set the `GNUPGHOME` environment variable.

Given that my setup means that I'm ( by default ) invoking a binary - `kata-agent` - which doesn't run as a Bash script, the current setup - setting `GNUPGHOME` in `~/.bashrc` ain't gonna cut it ....

Watch this space ....

**Comment by mtrmac** [comment]
@hamburgerbun @davidhay1969 Were you able to diagnose the root cause?

**Comment by vrothberg** [comment]
Closing due to inactivity. Please let us know if you find time to get back to the issue.
